Documentation/design-docs/host-startup-hook.md

```diff
@@ -0,0 +1,93 @@
+# Host startup hook
+
+For .NET Core 3+, we want to provide a hook that allows managed code
+to run before the main application's entry point.
+
+## Motivation
+
+This would allow hosting providers to put configuration and policy in
+a single location, including settings that potentially influence load
+behavior of the main entry point such as the AssemblyLoadContext
+behavior. This could be used to set up callbacks for handling
+Debug.Assert (if we make such an API available), or other
+environment-dependent behavior. The hook needs to be separate from the
+entry point, so that user code doesn't need to be modified.
+
+## Proposed behavior
+
+Environment variables or `<appname>.runtimeconfig.json` can be used to
+specify a managed assembly and type that contains an `Initialize`
```

Comment:
Please define the exact signature in either C# (e.g. `public void Initialize()`)